### ISSUE TITLE
Bug for merge_goto(cur_state->guard)?

### ISSUE THREAD

**Issue by zhuyutian57** [esbmc/esbmc#1981]
In `merge_goto`, it seems that the `cur->state.guard` is updated in the wrong place. See below codes.

```c++
void goto_symext::merge_gotos()
{
  ...
  for (auto list_it = state_list.rbegin(); list_it != state_list.rend();
       list_it++)
  {
    statet::goto_statet &goto_state = *list_it;

    // Merge guards. Don't write this to `state` yet because we might move
    // goto_state over it below.
    guardt new_guard = merge_state_guards(goto_state, *cur_state);

    if (!goto_state.guard.is_false())
    {
      // do SSA phi functions
      phi_function(goto_state);

      merge_locality(goto_state);

      merge_value_sets(goto_state);

      // adjust depth
      cur_state->num_instructions =
        std::min(cur_state->num_instructions, goto_state.num_instructions);
    }

    cur_state->guard = std::move(new_guard);
  }

  // clean up to save some memory
  frame.goto_state_map.erase(state_map_it);
}
```

When calling `merge_state_guards`, the comment says `Don't write to 'state'`. Does it mean that `cur_state->guard` stay unchanged? However, inside `merge_state_guards`, `cur_state->guard` may be updated. See below codes.

```c++
static inline guardt merge_state_guards(
  goto_symext::statet::goto_statet &goto_state,
  goto_symex_statet &state)
{
  ...
  if (
    (!goto_state.guard.is_false() && !state.guard.is_false()) ||
    state.guard.disjunction_may_simplify(goto_state.guard))
  {
    state.guard |= goto_state.guard; // updated here, state is a reference.
    return state.guard;
  }
  else if (state.guard.is_false() && !goto_state.guard.is_false())
  {
    return goto_state.guard;
  }
  else
  {
    return state.guard;
  }
}
```

`state.guard |= goto_state.guard` will update `cur_state->guard`. Is this a bug? I think that `return state.guard | goto_state.guard` is right.

If we merging guards in this way, the result for merging value sets is wrong for the below example.

```c++
#include <stdlib.h>
extern _Bool __VERIFIER_nondet_bool();
int **g = ((void *)0);
int main() {
 g = (int **) malloc(sizeof(int *));
 *g = (int *) malloc(sizeof(int));
 if (__VERIFIER_nondet_bool()) {
     free(g);
     g = ((void *)0);
     if (g != ((void *)0)) {
          free(*g);
     }
     return 0;
    // pc_true_end
 } // pc_false
// pc_merged
 free(*g);
 free(g);
 g = ((void *)0);
 return 0;
}
```

Firstly, **--force-malloc--success** is open.

In the program point `// pc_merged`, we merge two states `// pc_false` and `// pc_true_end`. Ideally, since the program has exited in  `// pc_true_end`, the information of `// pc_merged` is equivalent to `// pc_false`. Right?

However, things does not go in this way because of the updating of `cur_state->guard` when constructing new guard.

Note that after `symex return` in `// pc_true_end`, the `cur_state` is the state in program point `// pc_true_end`. And the guard of becomes `false`. Moreover, the value set of `g` is empty(or null object).

When merging state `// pc_false` into the `cur_state`, the following code executes.
```c++
static inline guardt merge_state_guards(
  goto_symext::statet::goto_statet &goto_state,
  goto_symex_statet &state)
{
  ...
  if (
    (!goto_state.guard.is_false() && !state.guard.is_false()) ||
    state.guard.disjunction_may_simplify(goto_state.guard)) // true
  {
    state.guard |= goto_state.guard; // updated here, state is a reference.
    return state.guard;
  }
  ...
}
```
`goto_state` is `// pc_false` with guard `not __VERIFIER_nondet_bool` and `state` is `cur_state` with guard `false`. However, `state.guard.disjunction_may_simplify(goto_state.guard)` is `true` by the following logic.

```c++
bool guardt::disjunction_may_simplify(const guardt &other_guard) const
{
  if (
    is_true() || is_false() || other_guard.is_true() || other_guard.is_false())
    return true; // is_false() holds, return true
  ...
}
```

As a result, `state.guard |= goto_state.guard;` updates `cur_state_>guard`. 

So, if we merge value sets(or others), `cur_state->guard` is not `false`. Then, the information of `// pc_true_end` is not being dicarded. And we get that the value set of `g` contains two items, one of them is `null object`. I think it is a bug here.

One more question, why should we merge guard before merging others? It seems unreasonable.

**Comment by rafaelsamenezes** [comment]
I am not very familiar with this part of ESBMC. But from my understanding:

1. `When calling merge_state_guards, the comment says Don't write to 'state'. Does it mean that cur_state->guard stay unchanged`. Indeed, it seems wrong, maybe we should convert the function parameter to const ref instead.
2. `One more question, why should we merge guard before merging others? It seems unreasonable.` I think you are right, it shouldn't. Specially because the merge functions actually use the current state guard.

**Comment by fbrausse** [comment]
I've looked at this particalur piece of symex in the past while figuring out how to fix UB in guard operators in #1297. Due to this issue I rebased those changes and re-ran the SV-COMP benchmarks in https://github.com/esbmc/esbmc/actions/runs/10400589431 It fixes the guard merge order in addition to documentation changes and avoiding the UB in operator|=().

There are no crucial changes to master at https://github.com/esbmc/esbmc/actions/runs/10383689091

However, the wrong value set for `g` I couldn't reproduce - or maybe I missed them. Are they shown when running with `--show-symex-value-set`?

**Comment by zhuyutian57** [comment]
Ok. I get it.

The value set for `g` is replaced in `phi_function` correctly.

However, the guard is generated by getting diff between `cur_state->guard` and `goto_state.guard`

```c++
void goto_symext::phi_function(const statet::goto_statet &goto_state)
{
  ...
  guardt tmp_guard;
  if (
    !variables.empty() && !cur_state->guard.is_false() &&
    !goto_state.guard.is_false())
  {
    tmp_guard = goto_state.guard;

    // this gets the diff between the guards
    tmp_guard -= cur_state->guard;
  }
  ...
}
```

Also, is it ok to merge guard after merging others?
